### PR TITLE
Pwned Passwords Notifications

### PR DESCRIPTION
**Pull request by jonathanjuursema**
Adds Pwned Passwords functionality using the [Pwned Passwords API](https://haveibeenpwned.com/Passwords) (including k-anonymity feature).

* You can no longer change your password to a password that is present in a known data breach.
* If you are currently using an exposed password, you will receive an e-mail after you've logged in with that password. That e-mail will only be sent ONCE per user and exposed password combination.

There will be an [accompanying wiki page](https://wiki.proto.utwente.nl/ict/pwned-passwords) with more info and security implications.

### CODE REVIEW

File: app/Http/Controllers/AuthController.php [hunk 1]
```diff
@@ -572,6 +578,10 @@ public static function verifyCredentials($username, $password)
         }
 
         if ($user != null && Hash::check($password, $user->password)) {
+            if ((new PwnedPasswords())->setPassword($password)->isPwnedPassword() && HashMapItem::where('key', 'pwned-pass')->where('subkey', $user->id)->first() === null) {
```
Comment (thijsdortmann) [inline app/Http/Controllers/AuthController.php]: For efficiency, maybe do the HaspMapItem check first before doing the isPwnedPassword-check, as we can do the HasMapItem check locally.

File: app/Mail/PwnedPasswordNotification.php [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+    /**
+     * Build the message.
+     *
+     * @return $this
+     */
+    public function build()
+    {
+        return $this
+            ->subject('IMPORTANT SECURITY NOTICE: The password you use with S.A. Proto was exposed in a data breach.')
```
Comment (thijsdortmann) [inline app/Mail/PwnedPasswordNotification.php]: Maybe change the subject to something that immediately makes clear that this wasn't a data breach of Proto's data.

File: app/Http/Controllers/AuthController.php [hunk 3]
```diff
@@ -397,6 +400,9 @@ public function passwordChangePost(Request $request)
             } elseif (strlen($pass_new1) < 10) {
                 $request->session()->flash('flash_message', 'Your new password should be at least 10 characters long.');
                 return view('auth.passchange');
+            } elseif ((new PwnedPasswords())->setPassword($pass_new1)->isPwnedPassword()) {
```
Comment (thijsdortmann) [inline app/Http/Controllers/AuthController.php]: I haven't been able to validate this with a quick Google: does this library automatically use the k-anonimity feature? As I don't see a configuration parameter for this specified anywhere.

Comment (jonathanjuursema) [reply app/Http/Controllers/AuthController.php]: [Yes it does!](https://github.com/nickurt/laravel-pwned-passwords/blob/master/src/PwnedPasswords.php#L86)

File: app/Http/Controllers/AuthController.php (hunk 1 above)
Comment (jonathanjuursema) [reply app/Http/Controllers/AuthController.php]: Agreed.

File: app/Mail/PwnedPasswordNotification.php (hunk 2 above)
Comment (jonathanjuursema) [reply app/Mail/PwnedPasswordNotification.php]: I'm not sure if that is the right approach. It is stated very clearly in the e-mail, and perhaps this subject is a better trigger to read the e-mail.

Comment (thijsdortmann) [reply app/Mail/PwnedPasswordNotification.php]: Good point, agreed :)

**Pull request merged**